crates/sui-core/src/execution_engine.rs

```diff
@@ -123,6 +123,14 @@ fn execute_transaction<S: BackingPackageStore + ParentSync>(
         // TODO: Since we require all mutable objects to not show up more than
         // once across single tx, we should be able to run them in parallel.
         for single_tx in transaction_data.kind.into_single_transactions() {
+            if single_tx.uses_vm() {
+                // Charge gas for this VM execution
+                if let Err(e) = gas_status.charge_vm_gas() {
+                    result = Err(e);
+                    break;
+                }
+            }
```

Comment:
Why not just move this into the match arms for the items that call the VM? Are we also flat feeing publishing with the same flat fee? I imagine we don't want to but it looks like we are here?

Reply:
> Why not just move this into the match arms for the items that call the VM?

Because code duplication. I also think it's clearer to flag that we're charging for VM use.

> Are we also flat feeing publishing with the same flat fee? I imagine we don't want to but it looks like we are here?

Publishing can invoke VM as part of init